File: src/main/java/com/joecollins/models/general/PartyResult.java
package com.joecollins.models.general;

import com.joecollins.graphics.utils.ColorUtils;
import java.awt.Color;

public class PartyResult {

  public static PartyResult NO_RESULT = PartyResult.leading(null);

  private final Party party;
  private final boolean elected;

  public PartyResult(Party party, boolean elected) {
    this.party = party;
    this.elected = elected;
  }

  public static PartyResult elected(Party party) {
    return new PartyResult(party, true);
  }

  public static PartyResult leading(Party party) {
    return new PartyResult(party, false);
  }

  public Party getParty() {
    return party;
  }

  public boolean isElected() {
    return elected;
  }

  public Color getColor() {
    if (party == null) {
      return Color.BLACK;
    }
    if (elected) {
      return party.getColor();
    }
    return ColorUtils.lighten(party.getColor());
  }
}

File: src/main/kotlin/com/joecollins/models/general/PartyResult.kt
package com.joecollins.models.general

import com.joecollins.graphics.utils.ColorUtils
import java.awt.Color

data class PartyResult(val party: Party, val elected: Boolean) {

    val isElected = elected

    val winner = party.takeIf { elected }

    companion object {

        val TIE = PartyResult(Party.TIE, false)

        fun elected(party: Party?) = if (party == null) null else PartyResult(party, true)

        fun leading(party: Party?) = if (party == null) null else PartyResult(party, false)

        val PartyResult?.color: Color get() {
            return when {
                this == null -> Color.BLACK
                this.elected -> party.color
                else -> ColorUtils.lighten(party.color)
            }
        }
    }
}
